`Contract` Monad, custom `Prelude` and  `FromData`/`ToData` instances

vvtran opened the pull request:
- https://github.com/Plutonomicon/cardano-browser-tx/issues/178
- https://github.com/Plutonomicon/cardano-browser-tx/issues/154
- `Address` (and related addresses), `Value`, `TransactionInput`, `TransactionOutput`, `Maybe`, `Either` and others `FromData`/`ToData` instances.
- Added `minus` to `Natural` as it's quite widely used, with infix operator `^-` inline with the `AdditiveHemigroup` which we don't have. Furthermore, `Natural` now has a `Semiring` instance which should be helpful.
- I added some comments on `FromData` and `ToData` of `Ratio`, our current implementation accepts zero denominator, which we may need to provide safety for, either for the instances or construction.
- Created a type called `NatRatio` as a newtype over `Rational` inline with `plutus-extra`, which smart constructors.
- Various helpers in `Contract.Monad` for throwing and running the contract.
- `mconcat` + `Helpers` inside `Contract.Prelude` 
- `ScriptLookups` now has a `Maybe` and non-`Maybe` context version to help with those using `<>` or `foldMap` etc, it means the user doesn't need to write `Just scriptlookup1 <> Just scriptlookup2  <> scriptlookup3`. Furthermore there are some `unsafe` versions.
- Improved some of the helpers in `Scripts`.
- Deleted `Serialization/Bech32` because it's unused.
- I originally made the change for the `Interval` fields  `from` -> `ivFrom` but have since reverted to `from`, because our API is quite divergent from Plutus - there's no point using their name if we have row polymorphism, and it's a minor change when converting off-chain code.

ngua commented on src/Contract/Prelude.purs:
```diff
@@ -0,0 +1,302 @@
+-- | A custom Prelude that re-exports Purescript's prelude and further expands.
```
Maybe we could add an adaptation of `mconcat` since it's so common in off-chain code

ngua commented on src/Contract/Monad.purs:
```diff
@@ -0,0 +1,77 @@
+-- using `Contract (ExceptT ContractError QueryM)`? Or would we rather `Error`
+-- to make use of Javascript native erroring.
+derive newtype instance MonadThrow Error Contract
+derive newtype instance MonadError Error Contract
```
I think this is OK for now. If we do something like `ContractError`, it will be really restrictive for users who will want to use their own types. Once we take care of #107, we can revisit this

ngua commented on src/Contract/PlutusData.purs:
```diff
@@ -0,0 +1,84 @@
+-- | This module that defines query functionality via Ogmios to get `PlutusData`
```
Should we export `To`/`FromData` from here?

vvtran commented:
Ah, I forgot to newtype the `QueryConfig`, will do this first thing tomorrow.

vvtran commented:
Also, before I forget, I think I will export `Aff` and `Effect` from our prelude (without the helpers), in the same way Haskell has `IO` automatically.

ngua commented on src/Contract/Monad.purs:
```diff
@@ -0,0 +1,72 @@
+derive newtype instance MonadRec Contract
+
+-- | Throws an `Error` for any showable error.
+throwError :: forall (e :: Type). Show e => e -> Contract e
```
Should we maybe rename this since it shadows the `MonadThrow` effect? I'm not sure how important that is. We could call it `throwContractError` for clarity perhaps?

ngua replied:
Also I'm not sure how pleasant this would be to use in practice. The type signature for `MonadThrow`'s `throwError` is `forall a. e -> m a`, so at least we can get an `a` back from it. Here we get the `e`, which I'm not sure is what we want

ngua commented on src/Contract/Transaction.purs:
```diff
@@ -0,0 +1,153 @@
+import Data.Maybe (Maybe)
+import Data.Newtype (wrap)
+import QueryM (signTransaction, submitTransaction) as QueryM
+import Types.JsonWsp (OgmiosTxOut, OgmiosTxOutRef) as JsonWsp --  Perhaps it's best not to expose `JsonWsp`.
```
I agree with the comment

vvtran replied:
Hmm, actually, because I followed the chain index convention, `unspentOutputs` is currently using `Map TxOutRef OgmiosTxOut` as input so we'd need to expose this for now. I did think about changing this recently, generally script lookups will need refining as mentioned before.

vvtran replied:
https://github.com/Plutonomicon/cardano-browser-tx/issues/200

ngua commented on src/Contract/Transaction.purs:
```diff
@@ -0,0 +1,153 @@
+      )
+  , CostModel(CostModel)
+  , Costmdls(Costmdls)
+  -- , DataHash(DataHash)
```
Why is this commented out?

vvtran replied:
I tried to keep things somewhat modular (although not always admittedly), so the commented out things are address or data related. Since it's part of the transaction, I can add them back in.

ngua commented on src/Serialization/Address.purs:
```diff
@@ -171,6 +174,16 @@ instance Eq Address where
 instance Ord Address where
   compare = compare `on` addressBytes
 
+-- Plutus uses `PlutusTx.makeIsDataIndexed ''Address [('Address,0)]` on their
```
You could add the issue URL here

ngua commented on src/Contract/Monad.purs:
```diff
@@ -0,0 +1,72 @@
+-- | A module defining the `Contract` monad.
```
Can we define a `defaultQueryConfig` here (it needs to be effectful since it initializes websockets, in `Aff` not `Contract`)?

ngua reviewed (commented):
I think we also need `Contract.Wallet`

ngua commented on src/Contract/Scripts.purs:
```diff
@@ -0,0 +1,61 @@
+  , Validator(Validator)
+  , ValidatorHash(ValidatorHash)
+  ) as TypesScripts
+import Types.TypedValidator
```
What about `DatumType` and `RedeemerType`

ngua reviewed (commented):
Forgot something from my previous review

ngua commented on src/Contract/ScriptLookups.purs:
```diff
@@ -0,0 +1,30 @@
+  , ScriptLookups(ScriptLookups)
+  , generalise
+  , mintingPolicyM
+  , mkUnbalancedTx
```
Don't we need to wrap this in `Contract`

ngua replied:
I mean `mkUnbalancedTx`

vvtran replied:
Ah, my bad, will do.

ngua commented on src/Types/ScriptLookups.purs:
```diff
@@ -2,15 +2,19 @@ module Types.ScriptLookups
   ( MkUnbalancedTxError(..)
```
I'm OK with the solution you have for this, but I wanted to ask what you think of also exporting some `unsafe...` versions of lookups that need to call `validatorHash`, etc...?

vvtran replied:
It ultimately boils to how unsafe 
```
exports._scriptHashFromBytesImpl = maybe => bytes => {
    return hashFromImpl(CardanoWasm.ScriptHash.from_bytes)(maybe)(bytes);
};
```
is. Is this related to freeing allocated CSL objects? I guess it would just error out if it doesn't work right so it should be okay?

klntsky commented on src/Contract/Prelude.purs:
```diff
@@ -0,0 +1,324 @@
+
+mconcat
+  :: forall (f :: Type -> Type) (m :: Type). Foldable f => Monoid m => f m -> m
+mconcat = foldr (<>) mempty
```
`mconcat = Data.Foldable.fold`

vvtran replied:
The output of `Data.Foldable.fold` is very much in Purescript style, although I suspect the Haskellers will a return value as opposed to `Fold { step :: a -> Fold a b, finish :: Unit -> b }`, unless I've misunderstood? @ngua any thoughts on this?

vvtran replied:
Also, this is a left fold on a right associative operator `(<>)`? Why is why I used `foldr`. Hmm.

ngua replied:
No, I meant like Haskell's `mconcat`. I think we should return the monoidal value, not a `Fold`

klntsky commented on src/Types/NatRatio.purs:
```diff
@@ -0,0 +1,78 @@
+
+-- | `NatRatio` is a newtype over `Rational` with smart constructors to create a
+-- | `NatRatio` safely. Therefore, the data constructor is not exported.
+newtype NatRatio = NatRatio Rational
```
Can we remove Rational and do

```
newtype Rational = Rational (Ratio BigInt)
```

?

Rational isn't going to be used much, right?

vvtran replied:
I was told that it's used a lot, I'd need to investigate how the Plutus Rational is implemented and probably replicate that, regardless of zero denoms unfortunately. Also, I believe they have a type alias too.

klntsky replied:
Sorry, I meant the old `type Rational = Ratio BigInt` isn't going to be used because the intention is to use the safe newtype.

So let's remove the type synonym in order not to occupy a good name.

vvtran replied:
So `NatRatio` is meant to be different to `Rational`, not a replacement, `NatRatio` is just the "non-negative version" so we do need both. Although, perhaps a `Newtype` does make sense for `Rational`. This issue is the default behaviour of Purescript `Ratio` may different to Haskell's `Ratio`, in which case a newtype would make sense to replicate Haskell's behaviour, and therefore Pltuus.

klntsky commented on src/FromData.purs:
```diff
@@ -61,13 +85,20 @@ instance FromData ByteArray where
   fromData = Just
 
+-- | This covers `Bech32` which is just a type alias for `String`
+instance FromData String where
```
I'm against this, the user might expect it to work with any string. We only have bytestring instance in plutus as far as I remember, let's keep it that way.

vvtran replied:
That's fair, I was a bit unsure on this TBH. This will affect many other things including `PaymentPubKey` but not `PaymentPubKeyHash` which *I think* is the important one.

vvtran replied:
Removed BTW.

klntsky commented on src/FromData.purs:
```diff
@@ -16,22 +16,42 @@ import Data.Ratio (Ratio, reduce)
+  fromData _ = Nothing
+
+instance (FromData a, FromData b) => FromData (Either a b) where
+  fromData (Constr n [ pd ]) = case fromData pd of
```
Does it call itself?

vvtran replied:
Not sure what you mean? The `pd` is already a `PlutusData` by construction, so we just need to check where it is zero- or one-indexed, and whether it's a `Left` or `Right` inside the `Just` (where the `Just` comes from the `fromData` definition)

klntsky replied:
I believe this code is not going to work. I added a test and it failed.

Could you please take a look at https://github.com/Plutonomicon/cardano-browser-tx/tree/klntsky/either-fromdata-bug ?

Hopefully I'm not misunderstanding something.

vvtran replied:
@klntsky really good spot. I just realised I was overnesting the Just/Left, we just need one layer of Just to confirm a "successful" fromData. Where `one` or `zero` just represents Right or Left. Check out my latest push, this should work with your test hopefully.

vvtran replied:
Also note I can't write
```
case fromData pd of
    Just e | n == one -> Just (Left e)
    Just x | n == zero -> Just (Right x)
 ```
 because this concretises the type of `a` to be the same as `b`, hence
 ```

instance (FromData a, FromData b) => FromData (Either a b) where
  fromData (Constr n [ pd ])
    | n == zero = maybe Nothing (Just <<< Left) (fromData pd)
    | n == one = maybe Nothing (Just <<< Right) (fromData pd)
  fromData _ = Nothing
 ```
 I also fixed `Maybe`, there's only one type here, so we don't need to worry about any type confusion

vvtran replied:
Here's `Maybe` for your convenience:
```
instance FromData a => FromData (Maybe a) where
  fromData (Constr n [ pd ]) = case fromData pd of
    Just _ | n == one -> Just Nothing
    Just x | n == zero -> Just (Just x) -- Just is zero-indexed by Plutus
    _ -> Nothing
  fromData _ = Nothing
  ```

vvtran replied:
Just to be overly explicit, if your example, toData gives `Constr zero [ Integer one ]` and mapping the `Integer` in the array should give `Just one`, not `Just (Left one)` like I originally had.

ngua commented on src/Contract/Prelude.purs:
```diff
@@ -0,0 +1,324 @@
+-- | A custom Prelude that re-exports Purescript's prelude and further expands.
```
Maybe we can export `Generic` and `genericShow` as well

vvtran commented:
I'm going to export `unwrap`, `wrap` and `over` too.

vvtran commented:
Okay, tried to apply all suggestions
- Added `class Newtype` too to prelude.
- newtype derived `MonadAff` for `Contract` which should be helpful.
- Removed `FromData`/`ToData` for `String` and related, which affects `PublicKey` which I think is okay.

klntsky reviewed (dismissed):
See https://github.com/Plutonomicon/cardano-browser-tx/pull/185#discussion_r830244712

klntsky commented on src/ToData.purs:
```diff
@@ -49,6 +67,8 @@ instance (ToData k, ToData v) => ToData (Map k v) where
     where
     entries = Map.toUnfoldable mp :: Array (k /\ v)
 
+-- Note that nothing prevents the denominator from being zero, we could provide
```
It looks like the property of not having zero denominator is not an invariant of `Ratio`, so we shouldn't do that. (Why though? I don't know)

I think we can remove the comment.

vvtran replied:
So it does seem like `Ratio` doesn't care about this property, and it's down to the user to do whatever they wish, in which case a `Newtype` as you previously suggested could be best. I think the main thing is to replicate whatever Plutus does

vvtran replied:
Happy to leave this for a separate PR btw. Just to get this merged in asap.

klntsky commented on src/FromData.purs:
```diff
@@ -41,16 +41,15 @@ instance FromData Boolean where
   fromData (Constr n [ pd ]) = case fromData pd of
-    Just Nothing | n == one -> Just Nothing
-    Just (Just x) | n == zero -> Just (Just x) -- Just is zero-indexed by Plutus
+    Just _ | n == one -> Just Nothing
```
Wait, you are still pattern match with `[ pd ]`, but in case of `Nothing` it's `[]`:

```
> toData (Nothing :: Maybe Integer)
Constr 1 []
```

vvtran replied:
Ah yes, you're correct, my bad, I  just changed it to:
```
instance FromData a => FromData (Maybe a) where
  fromData (Constr n [ pd ])
    | n == zero = maybe Nothing (Just <<< Just) (fromData pd) -- Just is zero-indexed by Plutus
  fromData (Constr n [])
    | n == one = Just Nothing
  fromData _ = Nothing  ```

ngua commented on src/Contract/Monad.purs:
```diff
@@ -0,0 +1,137 @@
+    }
+
+-- | Same as `defaultContractConfigAff` but lifted into `Contract`.
+defaultContractConfig :: Contract ContractConfig
```
I'm not sure how useful this will be, since we need a `ContractConfig` to run any `Contract` action. But I suppose we can keep it (maybe it would be useful with `local` or something?)

vvtran replied:
It could be useful for local, although I wonder if `defaultContractConfig` is a better name for the `Aff` version since that's what everyone is using. Then call the `Contract` version `defaultContractConfigLifted` perhaps.

ngua replied:
I think that's the best/clearest way to name them
